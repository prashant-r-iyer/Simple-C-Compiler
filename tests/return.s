	.text

.globl compute
compute:
	# Save Frame pointer
	pushq %rbp
	movq %rsp, %rbp
	subq $256, %rsp
	pushq %rbx
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15
	pushq %r11
	pushq %r9
	movq %rdi, -8(%rbp)
	movq %rsi, -16(%rbp)
	movq -8(%rbp), %rbx
	movq -16(%rbp), %r10

	# +
	addq %r10, %rbx
	movq %rbx, -24(%rbp)
	movq -24(%rbp), %rbx
	movq %rbx, %rax

	# Restore registers
	popq %r9
	popq %r11
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	popq %rbx
	addq $256, %rsp

	leave
	ret


	# Restore registers
	popq %r9
	popq %r11
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	popq %rbx
	addq $256, %rsp

	leave
	ret

	.text

.globl main
main:
	# Save Frame pointer
	pushq %rbp
	movq %rsp, %rbp
	subq $256, %rsp
	pushq %rbx
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15
	pushq %r11
	pushq %r9

	# push 5
	movq $5, %rbx

	# push 6
	movq $6, %r10
	# func=compute nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	call compute
	movq %rax, %rbx
	movq %rbx, -8(%rbp)
	movq $string0, %rbx
	movq -8(%rbp), %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx

	# Restore registers
	popq %r9
	popq %r11
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	popq %rbx
	addq $256, %rsp

	leave
	ret

string0:
	.string "sum=%d\n"

