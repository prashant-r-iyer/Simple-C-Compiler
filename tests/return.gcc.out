return.c: In function ‘main’:
return.c:12:9: warning: implicit declaration of function ‘printf’ [-Wimplicit-function-declaration]
   12 |         printf("sum=%d\n", f);
      |         ^~~~~~
return.c:1:1: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
  +++ |+#include <stdio.h>
    1 | 
return.c:12:9: warning: incompatible implicit declaration of built-in function ‘printf’ [-Wbuiltin-declaration-mismatch]
   12 |         printf("sum=%d\n", f);
      |         ^~~~~~
return.c:12:9: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
return.c:12:22: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
   12 |         printf("sum=%d\n", f);
      |                     ~^     ~
      |                      |     |
      |                      int   long int
      |                     %ld
